firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,504 @@
+        CHKERR(DMLabelHasPoint(lbl_owned, p, &is_owned))
+        CHKERR(DMLabelHasPoint(lbl_ghost, p, &is_ghost))
+        if not is_ghost and not is_owned:
+            CHKERR(DMLabelSetValue(lbl_core, p, 1))
+    CHKERR(DMLabelDestroyIndex(lbl_owned))
+    CHKERR(DMLabelDestroyIndex(lbl_ghost))
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def get_entity_classes(PETSc.DM swarm):
+    """Builds PyOP2 entity class offsets for all entity levels.
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    :arg swarm: The DMSwarm object encapsulating the mesh topology
+    """
+    cdef:
+        np.ndarray[PetscInt, ndim=2, mode="c"] entity_class_sizes
+        np.ndarray[PetscInt, mode="c"] eStart, eEnd
+        PetscInt depth, d, i, ci, class_size, start, end
+        const PetscInt *indices = NULL
+        PETSc.IS class_is
+
+    depth = 1 # by definition since a swarm is point cloud
+    entity_class_sizes = np.zeros((depth, 3), dtype=IntType)
+    eStart = np.zeros(depth, dtype=IntType)
+    eEnd = np.zeros(depth, dtype=IntType)
+    for d in range(depth):
+        start = 0 # by definition since a swarm is point cloud
+        CHKERR(DMSwarmGetLocalSize(swarm.dm, &end)) # by definition since a swarm is point cloud
+        eStart[d] = start
+        eEnd[d] = end
+
+    for i, op2class in enumerate([b"pyop2_core",
+                                  b"pyop2_owned",
+                                  b"pyop2_ghost"]):
+        class_is = swarm.getStratumIS(op2class, 1)
+        class_size = swarm.getStratumSize(op2class, 1)
+        if class_size > 0:
+            CHKERR(ISGetIndices(class_is.iset, &indices))
+            for ci in range(class_size):
+                for d in range(depth):
+                    if eStart[d] <= indices[ci] < eEnd[d]:
+                        entity_class_sizes[d, i] += 1
+                        break
+            CHKERR(ISRestoreIndices(class_is.iset, &indices))
+
+    # PyOP2 entity class indices are additive
+    for d in range(depth):
+        for i in range(1, 3):
+            entity_class_sizes[d, i] += entity_class_sizes[d, i-1]
+    return entity_class_sizes
+
+def create_section(mesh, nodes_per_entity, on_base=False):
+    """Create the section describing a global numbering.
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    :arg mesh: The mesh.
+    :arg nodes_per_entity: Number of nodes on each type of topological
+        entity of the mesh.
+    :arg on_base: If True, assume extruded space is actually Foo x Real.
+
+    :returns: A PETSc Section providing the number of dofs, and offset
+        of each dof, on each mesh point.
+    """
+    # We don't use DMPlexCreateSection because we only ever put one
+    # field in each section.
```

Comment:
comment irrelevant since we don't have a plex here.

Reply:
removed